File: backend/engine/threat_scorer.py
"""
Threat Scorer
Aggregates authentication results, NLP detection, link indicators, header anomalies,
IP geolocation risk, and domain reputation into an overall threat assessment.
"""

def calculate_threat_score(analysis_results: dict) -> dict:
    header_anomalies = analysis_results.get("headerAnomalies", [])
    auth_result = analysis_results.get("authResult", {})
    nlp_result = analysis_results.get("nlpResult", {})
    link_result = analysis_results.get("linkResult", {})
    ip_results = analysis_results.get("ipResults", [])
    domain_result = analysis_results.get("domainResult", {})

    scores = {
        "authentication": 100 - (auth_result.get("overallScore") or 50),
        "nlp": nlp_result.get("nlpScore") or 0,
        "links": link_result.get("overallScore") or 0,
        "headerAnomalies": min(100, len(header_anomalies) * 20),
        "ipRisk": _calculate_aggregate_ip_risk(ip_results),
        "domainRisk": 100 - (domain_result.get("reputationScore") or 50) if domain_result else 50,
    }

    weights = {
        "authentication": 0.20,
        "nlp": 0.30,
        "links": 0.15,
        "headerAnomalies": 0.15,
        "ipRisk": 0.10,
        "domainRisk": 0.10,
    }

    weighted_score = sum(scores[k] * weights[k] for k in weights)
    threat_score = round(min(100, weighted_score))

    classification = _classify_threat(threat_score, analysis_results)
    confidence = _calculate_confidence(scores)
    risk_level = _get_risk_level(threat_score)
    summary = _generate_threat_summary(threat_score, classification, scores)
    recommendation = _get_recommendation(threat_score, classification)
    all_findings = _collect_findings(analysis_results)

    return {
        "threatScore": threat_score,
        "classification": classification,
        "confidence": confidence,
        "scores": scores,
        "summary": summary,
        "findings": all_findings,
        "riskLevel": risk_level,
        "recommendation": recommendation,
    }

def _calculate_aggregate_ip_risk(ip_results: list) -> int:
    if not ip_results:
        return 20
    max_risk = 0
    for ip in ip_results:
        risk = 0
        if ip.get("isProxy"):
            risk += 40
        if ip.get("isHosting"):
            risk += 20
        for ind in ip.get("riskIndicators", []):
            sev = ind.get("severity")
            if sev == "critical":
                risk += 40
            elif sev == "high":
                risk += 25
            elif sev == "medium":
                risk += 10
        max_risk = max(max_risk, min(100, risk))
    return max_risk

def _classify_threat(score: int, results: dict) -> str:
    nlp = results.get("nlpResult", {})
    bec_score = nlp.get("bec", {}).get("score", 0)
    imp_score = nlp.get("impersonation", {}).get("score", 0)
    phish_score = nlp.get("phishing", {}).get("score", 0)

    if bec_score > 30 and score > 35:
        return "fraud"
    if imp_score > 30 and score > 30:
        return "impersonation"
    if phish_score > 40 and score > 35:
        return "phishing"
    if score >= 65:
        return "phishing"
    if score >= 35:
        return "suspicious"
    return "legitimate"

def _get_risk_level(score: int) -> str:
    if score >= 75:
        return "critical"
    if score >= 55:
        return "high"
    if score >= 35:
        return "medium"
    if score >= 15:
        return "low"
    return "minimal"

def _calculate_confidence(scores: dict) -> int:
    signals = list(scores.values())
    high_count = sum(1 for s in signals if s > 50)
    low_count = sum(1 for s in signals if s < 20)

    if high_count >= 4:
        return 95
    if high_count >= 3:
        return 85
    if low_count >= 4:
        return 90
    if high_count >= 2:
        return 75
    return 65

def _generate_threat_summary(score: int, classification: str, scores: dict) -> str:
    parts = []
    if classification == "fraud":
        parts.append("⚠️ This email exhibits strong indicators of Business Email Compromise (BEC) or financial fraud.")
    elif classification == "impersonation":
        parts.append("🎭 This email shows signs of sender impersonation — the claimed identity does not match the actual sender infrastructure.")
    elif classification == "phishing":
        parts.append("🎣 This email contains multiple phishing indicators including deceptive language and suspicious links.")
    elif classification == "suspicious":
        parts.append("🔍 This email has some suspicious characteristics that warrant further investigation.")
    else:
        parts.append("✅ This email appears legitimate based on available indicators.")

    if scores.get("authentication", 0) > 60:
        parts.append("Email failed critical authentication checks (SPF/DKIM/DMARC).")
    if scores.get("nlp", 0) > 50:
        parts.append("Content analysis detected coercive or deceptive phrasing.")
    if scores.get("links", 0) > 50:
        parts.append("Contains links with high-risk attributes.")

    return " ".join(parts)

def _get_recommendation(score: int, classification: str) -> str:
    if score >= 75:
        return "BLOCK & QUARANTINE: Highly malicious email. Do not interact with links, attachments, or sender."
    if score >= 55:
        return "FLAG & WARN: Suspicious email. Quarantine pending administrator review."
    if score >= 35:
        return "EXERCISE CAUTION: Moderate risk detected. Verify sender identity through out-of-band communication."
    return "DELIVER NORMALLY: Email passed standard verification checks."

def _collect_findings(results: dict) -> list:
    all_findings = []
    
    # Header anomalies
    for a in results.get("headerAnomalies", []):
        all_findings.append({
            "category": "Header Forensics",
            "severity": "danger" if a.get("severity") == "high" else "warning",
            "description": a.get("description", ""),
            "details": [str(a.get("details", {}))]
        })

    # NLP findings
    for f in results.get("nlpResult", {}).get("findings", []):
        all_findings.append(f)

    # Link findings
    for f in results.get("linkResult", {}).get("findings", []):
        all_findings.append(f)

    # Domain findings
    for f in results.get("domainResult", {}).get("findings", []):
        all_findings.append(f)

    return all_findings

File: frontend/src/engine/threatScorer.js
/**
 * Threat Scorer
 * Aggregates all analysis modules into a unified threat assessment.
 */

/**
 * Calculate comprehensive threat score and classification
 */
export function calculateThreatScore(analysisResults) {
  const {
    headerAnomalies = [],
    authResult = {},
    nlpResult = {},
    linkResult = {},
    ipResults = [],
    domainResult = {},
  } = analysisResults;

  // Weight each analysis component
  const scores = {
    authentication: 100 - (authResult.overallScore || 50),   // Invert: low auth = high threat
    nlp: nlpResult.nlpScore || 0,
    links: linkResult.overallScore || 0,
    headerAnomalies: Math.min(100, headerAnomalies.length * 20),
    ipRisk: calculateAggregateIPRisk(ipResults),
    domainRisk: 100 - (domainResult?.reputationScore || 50),  // Invert
  };

  const weights = {
    authentication: 0.20,
    nlp: 0.30,
    links: 0.15,
    headerAnomalies: 0.15,
    ipRisk: 0.10,
    domainRisk: 0.10,
  };

  let weightedScore = 0;
  for (const [key, weight] of Object.entries(weights)) {
    weightedScore += (scores[key] || 0) * weight;
  }

  const threatScore = Math.round(Math.min(100, weightedScore));

  // Determine classification
  const classification = classifyThreat(threatScore, analysisResults);

  // Generate summary
  const summary = generateThreatSummary(threatScore, classification, scores, analysisResults);

  // Collect all findings
  const allFindings = collectFindings(analysisResults);

  return {
    threatScore,
    classification,
    confidence: calculateConfidence(scores),
    scores,
    summary,
    findings: allFindings,
    riskLevel: getRiskLevel(threatScore),
    recommendation: getRecommendation(threatScore, classification),
  };
}

/**
 * Calculate aggregate IP risk from all IP results
 */
function calculateAggregateIPRisk(ipResults) {
  if (!ipResults || ipResults.length === 0) return 20;

  let maxRisk = 0;
  for (const ip of ipResults) {
    let risk = 0;
    if (ip.isProxy) risk += 40;
    if (ip.isHosting) risk += 20;
    for (const indicator of (ip.riskIndicators || [])) {
      switch (indicator.severity) {
        case 'critical': risk += 40; break;
        case 'high': risk += 25; break;
        case 'medium': risk += 10; break;
      }
    }
    maxRisk = Math.max(maxRisk, Math.min(100, risk));
  }

  return maxRisk;
}

/**
 * Classify the threat type
 */
function classifyThreat(score, results) {
  const nlp = results.nlpResult || {};
  const hasBEC = nlp.bec?.score > 30;
  const hasImpersonation = nlp.impersonation?.score > 30;
  const hasPhishing = nlp.phishing?.score > 40;

  if (hasBEC && score > 35) return 'fraud';
  if (hasImpersonation && score > 30) return 'impersonation';
  if (hasPhishing && score > 35) return 'phishing';
  if (score >= 65) return 'phishing';
  if (score >= 40) return 'suspicious';
  if (score >= 20) return 'suspicious';
  return 'legitimate';
}

/**
 * Get risk level string
 */
function getRiskLevel(score) {
  if (score >= 75) return 'critical';
  if (score >= 55) return 'high';
  if (score >= 35) return 'medium';
  if (score >= 15) return 'low';
  return 'minimal';
}

/**
 * Calculate confidence percentage
 */
function calculateConfidence(scores) {
  // Higher confidence when multiple signals agree
  const signals = Object.values(scores);
  const highSignals = signals.filter(s => s > 50).length;
  const lowSignals = signals.filter(s => s < 20).length;

  if (highSignals >= 4) return 95;
  if (highSignals >= 3) return 85;
  if (lowSignals >= 4) return 90; // Confident it's safe
  if (highSignals >= 2) return 75;

  return 65;
}

/**
 * Generate human-readable threat summary
 */
function generateThreatSummary(score, classification, scores, results) {
  const parts = [];

  switch (classification) {
    case 'fraud':
      parts.push('⚠️ This email exhibits strong indicators of Business Email Compromise (BEC) or financial fraud.');
      break;
    case 'impersonation':
      parts.push('🎭 This email shows signs of sender impersonation — the claimed identity does not match the actual sender infrastructure.');
      break;
    case 'phishing':
      parts.push('🎣 This email contains multiple phishing indicators including deceptive content and suspicious links.');
      break;
    case 'suspicious':
      parts.push('🔍 This email has some suspicious characteristics that warrant further investigation.');
      break;
    case 'legitimate':
      parts.push('✅ This email appears legitimate based on available indicators.');
      break;
  }

  // Add specific insights
  if (scores.authentication > 60) {
    parts.push('Email authentication (SPF/DKIM/DMARC) is failing or absent.');
  }

  if (scores.nlp > 50) {
    parts.push('Content analysis detected social engineering language patterns.');
  }

  if (scores.links > 40) {
    parts.push('Embedded links show high-risk characteristics.');
  }

  if (scores.headerAnomalies > 30) {
    parts.push('Header analysis revealed routing anomalies.');
  }

  return parts.join(' ');
}

/**
 * Collect all findings from all analysis modules
 */
function collectFindings(results) {
  const findings = [];

  // Header anomalies
  const anomalies = results.headerAnomalies || [];
  for (const anomaly of anomalies) {
    findings.push({
      category: 'Header Anomaly',
      severity: anomaly.severity === 'high' ? 'danger' : anomaly.severity === 'medium' ? 'warning' : 'info',
      description: anomaly.description,
      details: anomaly.details ? [JSON.stringify(anomaly.details)] : [],
    });
  }

  // NLP findings
  if (results.nlpResult?.findings) {
    findings.push(...results.nlpResult.findings);
  }

  // Link findings
  if (results.linkResult?.findings) {
    findings.push(...results.linkResult.findings);
  }

  // Domain findings
  if (results.domainResult?.findings) {
    findings.push(...results.domainResult.findings);
  }

  // IP findings
  for (const ip of (results.ipResults || [])) {
    for (const indicator of (ip.riskIndicators || [])) {
      findings.push({
        category: 'IP Intelligence',
        severity: indicator.severity === 'critical' ? 'danger' : indicator.severity === 'high' ? 'danger' : 'warning',
        description: `${ip.ip}: ${indicator.description}`,
        details: [`Location: ${ip.city}, ${ip.country}`, `ISP: ${ip.isp}`],
      });
    }
  }

  // Sort by severity
  const severityOrder = { danger: 0, warning: 1, info: 2 };
  findings.sort((a, b) => (severityOrder[a.severity] || 99) - (severityOrder[b.severity] || 99));

  return findings;
}

/**
 * Get recommendation based on threat assessment
 */
function getRecommendation(score, classification) {
  if (score >= 75) {
    return {
      action: 'BLOCK',
      urgency: 'immediate',
      text: 'Immediately quarantine this email. Do not allow user interaction. Initiate incident response procedures and preserve evidence for forensic analysis.',
      color: 'danger',
    };
  }

  if (score >= 55) {
    return {
      action: 'QUARANTINE',
      urgency: 'high',
      text: 'Quarantine this email and flag for analyst review. Do not deliver to end user until investigation is complete.',
      color: 'warning',
    };
  }

  if (score >= 35) {
    return {
      action: 'REVIEW',
      urgency: 'medium',
      text: 'Flag this email for manual review. Some indicators suggest suspicious activity but confidence is moderate.',
      color: 'warning',
    };
  }

  if (score >= 15) {
    return {
      action: 'MONITOR',
      urgency: 'low',
      text: 'Minor suspicious indicators detected. Allow delivery but log for future reference and correlation.',
      color: 'info',
    };
  }

  return {
    action: 'ALLOW',
    urgency: 'none',
    text: 'Email appears legitimate. No immediate action required.',
    color: 'success',
  };
}
